mesonpy/__init__.py

```diff
@@ -364,10 +375,23 @@ def _select_abi_tag(self) -> Optional[mesonpy._tags.Tag]:  # noqa: C901
                     )
         return selected_tag
 
-    def _is_elf(self, file: Union[str, pathlib.Path]) -> bool:
-        """Check if file is an ELF file."""
+    def _is_native(self, file: Union[str, pathlib.Path]) -> bool:
+        """Check if file is a native file."""
+        self._project.build()  # the project needs to be built for this :/
+
         with open(file, 'rb') as f:
-            return f.read(4) == b'\x7fELF'
+            if platform.system() == 'Linux':
+                return f.read(4) == b'\x7fELF'  # ELF
+            elif platform.system() == 'Darwin':
+                return f.read(4) in (
+                    b'\xfe\xed\xfa\xce',  # 32-bit
+                    b'\xfe\xed\xfa\xcf',  # 64-bit
+                    b'\xca\xfe\xba\xbe',  # universal / fat (same as java class so beware!)
+                )
+            elif platform.system() == 'Windows':
+                return f.read(2) == b'MZ'
+
+        raise NotImplementedError(f'Unknown system: {platform.system()}')
```

Comment:
You can't do this, that is guaranteed to break things on other platforms. I suggest emitting a warning and returning `True`, because that's the safest choice.